src/safepoints.rs

```diff
@@ -0,0 +1,138 @@
+extern crate ykstackmaps;
+
+use std::collections::HashMap;
+use std::path::Path;
+use ykstackmaps::{LocKind, LocOffset, SMRec, StackMapParser};
+
+use core::mem;
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct ReturnAddress(pub u64);
+
+/// The offset from the stack pointer.
+#[derive(Debug)]
+struct SPO(u32);
+
+/// A `PtrSlot` identifies a stack root at a given safepoint using its offset
+/// from the Stack Pointer.
+///
+/// A base pointer (not to be confused with X86 terminology, where base pointer
+/// refers to the frame pointer register) is a pointer an object. In opposition
+/// to this, a derived pointer points to the interior of an object.
+///
+/// The Derived variant of a `PtrSlot` also contains a Stack Pointer offset to
+/// the base of the object.
+#[derive(Debug)]
+enum PtrSlot {
+    Base(SPO),
+    Derived(SPO, SPO),
+}
+
+/// Contains root locations for a Safepoint
+#[derive(Debug)]
+pub struct SafepointRoots {
+    /// A list of registers which contain roots across a safepoint
+    // DWARF Register number mapping can be found here:
+    // Pg.63 https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf
+    registers: Vec<u16>,
```

Comment:
Is performance critical here? Since the set of possible registers is fixed (and small?), you could use a bitset?